bench.c

```diff
@@ -0,0 +1,189 @@
+ * (mainly caused by configs (kernel: `tcp_max_syn_backlog` and
+ * `somaxconn`. Application (fastecho): `backlog`). Nevertheless, default
+ * maximum number of fd per-process is `1024`, if you
+ * insist to proceed the benchmark, perform following modifications:
+ *       (PROCEDURE MAY VARY BETWEEN DIFFERENT DISTROS)
+ * For Ubuntu:
+ * (1)
+ * Use following commands to adjust kernel attributes:
+ *     a. `$ sudo sysctl net.core.somaxconn={depends_on_MAX_THREAD}`
+ *     b. `$ sudo sysctl net.ipv4.tcp_max_syn_backlog={ditto}`
+ * Note that `$ sudo sysctl net.core.somaxconn` can get current value.
+ * `somaxconn` is max amount of established connection, whereas
+ * `tcp_max_syn_backlog` is max amount of connection at first step
+ * of TCP 3-way handshake (SYN).
+ *
+ * (2)
+ * Use command `$ ulimit -n {ditto}` to enlarge limitation of fd
+ * per-process. Note that this modification only effects on process
+ * which executes the command and its child processes.
+ *
+ * (3)
+ * Specify `backlog` with value as large as `net.ipv4.tcp_max_syn_backlog`.
+ *
+ * Remember to reset the modifications after benchmarking to keep
+ * stability of running machine
+ */
+#define MAX_THREAD 1000
+
+/*
+ * TODO: use vary-length msg instead would be better since it's more
+ * conforms to reality
+ */
+const char *msg_dum = "dummy message";
+
+pthread_t pt[MAX_THREAD];
+
+/* block all worker before they are all ready to bombing kecho */
+bool ready;
+
+pthread_mutex_t res_lock = PTHREAD_MUTEX_INITIALIZER;
+pthread_mutex_t worker_lock = PTHREAD_MUTEX_INITIALIZER;
+pthread_cond_t worker_wait = PTHREAD_COND_INITIALIZER;
+
+long time_res[MAX_THREAD] = {0};
+int idx = 0; /* for indexing `time_res` */
+FILE *bench_fd;
+
+static inline long time_diff_us(struct timeval *start, struct timeval *end)
+{
+    return ((end->tv_sec - start->tv_sec) * 1000000) +
+           (end->tv_usec - start->tv_usec);
+}
+
+static void *bomb_worker(void *arg)
+{
+    int sock_fd, rt;
+    char dummy[MAX_MSG_LEN];
+    struct timeval start, end;
+
+    /* wait until all worker created */
+    pthread_mutex_lock(&worker_lock);
+    while (!ready)
+        if ((rt = pthread_cond_wait(&worker_wait, &worker_lock))) {
+            puts("cond_wait failed!");
+            exit(-1);
+        }
+    pthread_mutex_unlock(&worker_lock);
+
+    sock_fd = socket(AF_INET, SOCK_STREAM, 0);
+    if (sock_fd == -1) {
+        goto sock_init_fail;
+    }
+
+    struct sockaddr_in info = {
+        .sin_family = PF_INET,
+        .sin_addr.s_addr = inet_addr(TARGET_HOST),
+        .sin_port = htons(TARGET_PORT),
+    };
+
+    if (connect(sock_fd, (struct sockaddr *) &info, sizeof(info)) == -1) {
+        goto connect_fail;
+    }
+
+    gettimeofday(&start, NULL);
+    send(sock_fd, msg_dum, strlen(msg_dum), 0);
+    recv(sock_fd, dummy, MAX_MSG_LEN, 0);
+    gettimeofday(&end, NULL);
+
+    close(sock_fd);
+
+    pthread_mutex_lock(&res_lock);
+    time_res[idx++] += time_diff_us(&start, &end);
```

Comment:
Because the test may run multiple times. At completion of all works, it calculates average result for each element of `time_res`.

The calculation is performed at end of `bench()`:
```c
for (int i = 0; i < MAX_THREAD; i++)
        fprintf(bench_fd, "%d %ld\n", i, time_res[i] /= BENCH_COUNT);
```